gateway/gateway_test.go

```diff
@@ -39,3 +42,198 @@ func TestValidate(t *testing.T) {
+			newNode: &node.Node{
+				DecoderPath: "new/decoder/path",
+				NodeName:    "NewNode",
+				JoinType:    "OTAA",
+				AppKey:      testAppKey,
+				DevEui:      testDevEUI,
+				AppSKey: []byte{
+					0x11, 0x11, 0x40, 0x4C,
+					0x69, 0x6E, 0x6B, 0x4C,
+					0x6F, 0x52, 0x61, 0x32,
+					0x30, 0x31, 0x38, 0x23,
+				},
+			},
+			oldNode: &node.Node{
+				Addr:    testDeviceAddr,
+				AppSKey: testAppSKey,
+			},
+			want: &node.Node{
+				DecoderPath: "new/decoder/path",
+				NodeName:    "NewNode",
+				JoinType:    "OTAA",
+				Addr:        testDeviceAddr,
+				AppSKey:     testAppSKey,
+				AppKey:      testAppKey,
+				DevEui:      testDevEUI,
+			},
+			expectErr: nil,
+		},
+		{
+			name: "ABP join type - use new Addr and AppSKey",
+			newNode: &node.Node{
+				DecoderPath: "new/decoder/path",
+				NodeName:    "NewNode",
+				JoinType:    "ABP",
+				Addr:        testDeviceAddr,
+				AppSKey:     testAppSKey,
+			},
+			oldNode: &node.Node{
+				Addr: []byte{1, 5, 6, 7},
+				AppSKey: []byte{
+					0x11, 0x11, 0x40, 0x4C,
+					0x69, 0x6E, 0x6B, 0x4C,
+					0x6F, 0x52, 0x61, 0x32,
+					0x30, 0x31, 0x38, 0x23,
+				},
+			},
+			want: &node.Node{
+				DecoderPath: "new/decoder/path",
+				NodeName:    "NewNode",
+				JoinType:    "ABP",
+				Addr:        testDeviceAddr,
+				AppSKey:     testAppSKey,
+			},
+			expectErr: nil,
+		},
+		{
+			name: "Invalid join type",
+			newNode: &node.Node{
+				JoinType: "INVALID",
+			},
+			oldNode:   &node.Node{},
+			want:      nil,
+			expectErr: errUnexpectedJoinType,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := mergeNodes(tt.newNode, tt.oldNode)
+			if tt.expectErr != nil {
+				test.That(t, err, test.ShouldBeError, tt.expectErr)
+			} else {
+				test.That(t, err, test.ShouldBeNil)
+				test.That(t, got, test.ShouldResemble, tt.want)
+			}
+		})
+	}
+}
+
+func TestUpdateReadings(t *testing.T) {
+	g := &Gateway{
+		lastReadings: make(map[string]interface{}),
+	}
+
+	// Test adding new readings
+	newReadings := map[string]interface{}{
+		"temp": 25.5,
+		"hum":  60,
+	}
+	g.updateReadings("device1", newReadings)
+	test.That(t, g.lastReadings["device1"], test.ShouldResemble, newReadings)
+
+	// Test updating existing readings
+	updatedReadings := map[string]interface{}{
+		"temp": 26.5,
+		"pres": 1013,
+	}
+	g.updateReadings("device1", updatedReadings)
+	expected := map[string]interface{}{
+		"temp": 26.5,
+		"hum":  60,
+		"pres": 1013,
+	}
+	test.That(t, g.lastReadings["device1"], test.ShouldResemble, expected)
+}
+
+func TestReadings(t *testing.T) {
+	g := &Gateway{
+		lastReadings: map[string]interface{}{
+			"device1": map[string]interface{}{
+				"temp": 25.5,
+				"hum":  60,
+			},
+			"device2": map[string]interface{}{
+				"pres": 1013,
+			},
+		},
+	}
+
+	readings, err := g.Readings(context.Background(), nil)
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, readings, test.ShouldResemble, g.lastReadings)
+}
```

Comment:
Can we combine with the test above, this isn't super useful on it's own for testing any functionality. and updateReadings -> expected readings from the API call is a better test of [encapsulation](https://testmatick.com/encapsulation-as-one-of-the-fundamental-principles-of-object-oriented-programming/).